src/gui/qgspointcloudquerybuilder.cpp

```diff
@@ -0,0 +1,382 @@
+  mTxtSql->setCursorPosition( i, j - 2 );
+  mTxtSql->setFocus();
+}
+
+void QgsPointCloudQueryBuilder::btnNotIn_clicked()
+{
+  mTxtSql->insertText( QStringLiteral( "NOT IN () " ) );
+  int i, j;
+  mTxtSql->getCursorPosition( &i, &j );
+  mTxtSql->setCursorPosition( i, j - 2 );
+  mTxtSql->setFocus();
+}
+
+void QgsPointCloudQueryBuilder::btnLessEqual_clicked()
+{
+  mTxtSql->insertText( QStringLiteral( "<= " ) );
+  mTxtSql->setFocus();
+}
+
+void QgsPointCloudQueryBuilder::btnGreaterEqual_clicked()
+{
+  mTxtSql->insertText( QStringLiteral( ">= " ) );
+  mTxtSql->setFocus();
+}
+
+void QgsPointCloudQueryBuilder::btnNotEqual_clicked()
+{
+  mTxtSql->insertText( QStringLiteral( "!= " ) );
+  mTxtSql->setFocus();
+}
+
+void QgsPointCloudQueryBuilder::btnAnd_clicked()
+{
+  mTxtSql->insertText( QStringLiteral( "AND " ) );
+  mTxtSql->setFocus();
+}
+
+void QgsPointCloudQueryBuilder::btnOr_clicked()
+{
+  mTxtSql->insertText( QStringLiteral( "OR " ) );
+  mTxtSql->setFocus();
+}
+
+void QgsPointCloudQueryBuilder::accept()
+{
+  if ( mTxtSql->text() != mOrigSubsetString )
+  {
+    if ( !mLayer->setSubsetString( mTxtSql->text() ) )
+    {
+      QMessageBox::warning( this, tr( "Query Result" ), tr( "Error in query. The subset string could not be set." ) );
+      return;
+    }
+  }
+
+  QDialog::accept();
+}
+
+void QgsPointCloudQueryBuilder::reject()
+{
+  if ( mLayer->subsetString() != mOrigSubsetString )
+    mLayer->setSubsetString( mOrigSubsetString );
+
+  QDialog::reject();
+}
+
+void QgsPointCloudQueryBuilder::test()
+{
+  QgsPointCloudExpression expression( mTxtSql->text() );
+  if ( !expression.isValid() && !mTxtSql->text().isEmpty() )
+  {
+    QMessageBox::warning( this,
+                          tr( "Query Result" ),
+                          tr( "An error occurred while parsing the expression:\n%1" ).arg( expression.parserErrorString() ) );
+  }
+  else
+  {
+    const QSet<QString> attributes = expression.referencedAttributes();
+    int offset;
+    for ( const auto &attribute : attributes )
+    {
+      if ( mLayer->dataProvider() &&
+           !mLayer->dataProvider()->attributes().find( attribute, offset ) )
+      {
+        QMessageBox::warning( this,
+                              tr( "Query Result" ),
+                              tr( "\"%1\" not recognized as an available attribute." ).arg( attribute ) );
+        return;
+      }
+    }
+    mLayer->setSubsetString( mTxtSql->text() );
+    QMessageBox::information( this,
+                              tr( "Query Result" ),
+                              tr( "The expression was successfully parsed." ) );
+  }
+}
+
+void QgsPointCloudQueryBuilder::clear()
+{
+  mTxtSql->clear();
+  mLayer->setSubsetString( QString() );
+}
+
+void QgsPointCloudQueryBuilder::showHelp()
+{
+  // TODO: No pointcloud help page yet
+  //QgsHelp::openHelp( QStringLiteral( "working_with_vector/vector_properties.html#query-builder" ) );
+}
+
+void QgsPointCloudQueryBuilder::saveQuery()
```

Comment:
this is now the 3rd identical copy of the saveQuery/loadQuery functions (also in QgsQueryBuilder and QgsSearchQueryBuilder. Could the versions in QgsQueryBuilder be made static and called directly from here instead?

Reply:
Ok! Should I also update the deprecated QgsSearchQueryBuilder to use it too?

Reply:
> Should I also update the deprecated QgsSearchQueryBuilder to use it too?

Up to you... if it's trivial to do so, than I'd say yes.

Reply:
ok